File: 2015/day-6.js
const fs = require("fs");
const path = "input-6.txt";
// const path = "test.txt";
const arraySize = 1000;

function solve() {
  const instructions = cleanInstructions(path);
  const grid = new Array(arraySize);
  buildGrid(grid);

  // Follow instructions from input file
  instructions.forEach((instruction) => {
    const start = instruction["start"];
    const stop = instruction["stop"];

    for (let x = start[0]; x <= stop[0]; x++) {
      for (let y = start[1]; y <= stop[1]; y++) {
        grid[x][y] = adjustLight(grid[x][y], instruction["power"]);
      }
    }
  });

  // Determine how much power is used
  totalPower = 0;
  for (let x = 0; x < grid.length; x++) {
    for (let y = 0; y < grid[x].length; y++) {
      totalPower += grid[x][y];
    }
  }

  console.log("Total power used: ", totalPower);
}

// Convert instructions to an array of objects
function cleanInstructions(path) {
  const data = fs.readFileSync(path).toString().split("\r\n");
  const instructions = [];

  data.forEach((line) => {
    line = line.replace(" through ", " ");
    if (line.startsWith("turn ")) {
      line = line.replace(" ", "_");
    }
    line = line.split(" ");

    const instruction = {};
    instruction["power"] = line[0];
    const start = line[1].split(",");
    const stop = line[2].split(",");
    instruction["start"] = [parseInt(start[0]), parseInt(start[1])];
    instruction["stop"] = [parseInt(stop[0]), parseInt(stop[1])];
    instructions.push(instruction);
  });
  return instructions;
}

function buildGrid(grid) {
  for (let i = 0; i < arraySize; i++) {
    const row = new Array(arraySize);
    row.fill(0);
    grid[i] = row;
  }
}

function adjustLight(light, power) {
  if (power === "toggle") {
    light += 2;
  } else if (power === "turn_on") {
    light += 1;
  } else if (power === "turn_off") {
    if (light > 0) {
      light -= 1;
    }
  }
  return light;
}

solve();

File: 2015/day-6.py
import numpy as np

data_path = "input-6.txt"
grid_size = (1000, 1000)
# data_path = "test.txt"
# grid_size = (10, 10)


def clean_instructions(data_path):
    with open(data_path) as f:
        data = f.read().split("\n")

    instructions = []

    for line in data:
        line = line.replace(" through ", " ")
        if line.startswith("turn "):
            line = line.replace(" ", "_", 1)
        line = line.split(" ")

        instruction = {}
        instruction["power"] = line[0]
        start = line[1].split(",")
        stop = line[2].split(",")
        instruction["start"] = tuple([int(start[0]), int(start[1])])
        instruction["stop"] = tuple([int(stop[0]), int(stop[1])])
        instructions.append(instruction)

    return instructions


def turn_on_or_off(grid, x, y, power):
    if power == "toggle":
        if grid[x][y]:
            grid[x][y] = False
        else:
            grid[x][y] = True
    elif power == "turn_on":
        grid[x][y] = True
    elif power == "turn_off":
        grid[x][y] = False


def turn_up_or_down(grid, x, y, power):
    if power == "toggle":
        grid[x][y] += 2
    elif power == "turn_on":
        grid[x][y] += 1
    elif power == "turn_off":
        if grid[x][y] > 0:
            grid[x][y] -= 1


def solve():
    instructions = clean_instructions(data_path)
    grid = np.full(grid_size, 0, dtype=int)

    for instruction in instructions:
        start = instruction["start"]
        stop = instruction["stop"]

        for x in range(start[0], stop[0] + 1):
            for y in range(start[1], stop[1] + 1):
                # turn_on_or_off(grid, x, y, instruction["power"]) # <- q1 solution
                turn_up_or_down(grid, x, y, instruction["power"])  # <- q2 solution

    lights_on = 0

    for row in grid:
        for cell in row:
            lights_on += cell

    print("Lights on:", lights_on)


solve()
